receptor_affinity/mesh.py

```diff
@@ -427,6 +416,36 @@ def ping(self, count=10):
             results[node.name] = node.ping(count)
         return results
 
+    def settle(self):
+        """Wait for the mesh and its nodes to settle on the same routes.
+
+        If the nodes and mesh don't agree within approximately 30 seconds, throw
+        ``RouteMismatchError``.
+        """
+        # There's two reasons to catch TimedOutError. First, it doesn't give terribly useful
+        # diagnostic information, whereas RouteMismatchError (raised by validate_routes()) does.
+        # Second, we want this library to hide implementation details, like the fact that we depend
+        # on the wait-for package.
+        #
+        # It'd be nice if we could raise the original RouteMismatchError. Unfortunately,
+        # TimedOutError doesn't hold a reference to the underlying exception.
+        #
+        # We don't call validate_routes() for a final time from within the except: block, so as to
+        # make tracebacks more concise.
+        settled = True
+        try:
+            wait_for(
+                self.validate_routes,
+                delay=6,
+                num_sec=30,
+                handle_exception=True,
+                fail_condition=lambda result: isinstance(result, RouteMismatchError),
+            )
+        except TimedOutError:
+            settled = False
+        if not settled:
+            self.validate_routes()
```

Comment:
I'm presuming you are running it again to get the exception? Did you know that wait_for should return the output if you ask for a silent_fail, you could then check if it's an exception, as opposed to having to run it again?

Reply:
Yes, I'm running it again to get the exception. If there's a way to get at the original error, I'd much prefer to do that. I'll re-read the source code and try again. :smile:

Reply:
I don't think that setting `silent_fail` allows `wait_for()` to return the captured exception. I think that [this code](https://github.com/RedHatQE/wait_for/blob/master/wait_for/__init__.py#L181-L182):

```python
        except Exception as e:
            if handle_exception:
                out = fail_condition
```

...should be changed to this:

```python
        except Exception as e:
            if handle_exception:
                out = fail_condition
```